### DISCUSSION TITLE
NestJs: No exports found error when deploying multiple modules

### DISCUSSION THREAD

**Discussion by jonathanvd** [vercel/vercel#9199] [Help]
Hello,

I'm trying to get started deploying a NestJS API to Vercel, building an API for an online recipebook.

Example repo: [Github](https://github.com/jonathanvd/pantry-api-example)

I started from the basic boilerplate app NestJS generates for you and it worked fine, both locally and deployed on Vercel

But when I try to add a module for handling recipes, in this case containing just a simple POST to upload a new recipe, the build still succeeds but when I try to call any function (the POST or the basic hello world the default app provides) it gives me the following error:

`[Nest] 7  - 
01/10/2023, 8:58:04 PM 
    LOG
 
[NestFactory] 
Starting Nest application...
[Nest] 7  - 
01/10/2023, 8:58:04 PM 
  ERROR
 
[MongooseModule] 
Unable to connect to the database. Retrying (1)...
[Nest] 7  - 
01/10/2023, 8:58:04 PM 
    LOG
 
[InstanceLoader] 
MongooseModule dependencies initialized
 +2ms
[Nest] 7  - 
01/10/2023, 8:58:04 PM 
    LOG
 
[InstanceLoader] 
AppModule dependencies initialized
 +1ms
[Nest] 7  - 
01/10/2023, 8:58:07 PM 
  ERROR
 
[MongooseModule] 
Unable to connect to the database. Retrying (2)...
2023-01-10T20:58:09.233Z	cec5fa9c-7892-4b10-9ab3-de4f4fa2dba5	ERROR	No exports found in module "../src/main.js".
2023-01-10T20:58:09.233Z	cec5fa9c-7892-4b10-9ab3-de4f4fa2dba5	ERROR	Did you forget to export a function or a server?
RequestId: cec5fa9c-7892-4b10-9ab3-de4f4fa2dba5 Error: Runtime exited with error: exit status 1
Runtime.ExitError`

The unable to connect to the database is not related here since I just omitted my MongoDB URI in the repo for security reasons. The error below it still happens if I have a correct URI.

The build log in case anyone needs it:

<html>
<body>

Cloning github.com/jonathanvd/pantry-api (Branch: refactor/cleanup, Commit: 93c2164)
--
21:56:55.209 | Cloning completed: 998.714ms
21:56:56.186 | Restored build cache
21:56:56.241 | Running "vercel build"
21:56:56.907 | Vercel CLI 28.10.3
21:56:57.136 | WARN! Due to `builds` existing in your configuration file, the Build and Development Settings defined in your Project Settings will not apply. Learn More: https://vercel.link/unused-build-settings
21:56:57.774 | Installing dependencies...
21:57:00.131 |  
21:57:00.132 | up to date in 2s
21:57:00.132 |  
21:57:00.132 | 92 packages are looking for funding
21:57:00.132 | run `npm fund` for details
21:57:00.449 | Using TypeScript 4.9.4 (local user-provided)
21:57:12.905 | Build Completed in /vercel/output [15s]
21:57:18.938 | Generated build outputs:
21:57:18.938 | - Static files: 0
21:57:18.938 | - Serverless Functions: 1
21:57:18.938 | - Edge Functions: 0
21:57:18.939 | Serverless regions: Paris, France
21:57:18.939 | Deployed outputs in 4s
21:57:19.436 | Build completed. Populating build cache...
21:57:29.145 | Uploading build cache [21.73 MB]...
21:57:29.906 | Build cache uploaded: 760.738ms

</body>
</html>

I'm posting this here and not over at NestJS since it works completely fine when run locally, both in development and production mode. So there has to be something going wrong with the deployment process.

Hope I gave enough info for someone to maybe help me out here.

If you need any more info from me go ahead and ask.

Thanks!

**Comment by amyegan** [comment]
Thanks for reaching out @jonathanvd! My first thought it the `nest build` output might not be in a format that is fully compatible with the expected serverless environment structure.

Vercel deployments expect a certain format for output, specifically using [Build Output API (v3)](https://vercel.com/docs/build-output-api/v3#), which sets things up for Serverless, Edge, etc. Since I don't know how `nest build` works, I need to do some more digging before I know if that build output is compatible. 

Feel free to take a look through the documentation yourself if you're interested! You can find more Build Output API information at [vercel.com/docs/build-output-api/v3](https://vercel.com/docs/build-output-api/v3) and minimal code examples at [github.com/vercel/examples/tree/main/build-output-api](https://github.com/vercel/examples/tree/main/build-output-api)

**Reply by jonathanvd** [reply]
> Thanks for reaching out @jonathanvd! My first thought it the `nest build` output might not be in a format that is fully compatible with the expected serverless environment structure.
> 
> Vercel deployments expect a certain format for output, specifically using [Build Output API (v3)](https://vercel.com/docs/build-output-api/v3#), which sets things up for Serverless, Edge, etc. Since I don't know how `nest build` works, I need to do some more digging before I know if that build output is compatible.
> 
> Feel free to take a look through the documentation yourself if you're interested! You can find more Build Output API information at [vercel.com/docs/build-output-api/v3](https://vercel.com/docs/build-output-api/v3) and minimal code examples at [github.com/vercel/examples/tree/main/build-output-api](https://github.com/vercel/examples/tree/main/build-output-api)

Thanks! I'll have a look into the documentation.
